src/python/grpcio/grpc/_cython/_cygrpc/aio/callback_common.pyx.pxi

```diff
@@ -0,0 +1,113 @@
+        ))
+
+
+cdef class CallbackWrapper:
+
+    def __cinit__(self, object future, CallbackFailureHandler failure_handler):
+        self.context.functor.functor_run = self.functor_run
+        self.context.waiter = <cpython.PyObject*>future
+        self.context.failure_handler = <cpython.PyObject*>failure_handler
+        # NOTE(lidiz) Not using a list here, because this class is critical in
+        # data path. We should make it as efficient as possible.
+        self._reference_of_future = future
+        self._reference_of_failure_handler = failure_handler
+
+    @staticmethod
+    cdef void functor_run(
+            grpc_experimental_completion_queue_functor* functor,
+            int success):
+        cdef CallbackContext *context = <CallbackContext *>functor
+        if success == 0:
+            (<CallbackFailureHandler>context.failure_handler).handle(
+                <object>context.waiter)
+        else:
+            (<object>context.waiter).set_result(None)
+
+    cdef grpc_experimental_completion_queue_functor *c_functor(self):
+        return &self.context.functor
+
+
+cdef CallbackFailureHandler CQ_SHUTDOWN_FAILURE_HANDLER = CallbackFailureHandler(
+    'grpc_completion_queue_shutdown',
+    'Unknown',
+    RuntimeError)
+
+
+cdef class CallbackCompletionQueue:
+
+    def __cinit__(self):
+        self._shutdown_completed = asyncio.get_event_loop().create_future()
+        self._wrapper = CallbackWrapper(
+            self._shutdown_completed,
+            CQ_SHUTDOWN_FAILURE_HANDLER)
+        self._cq = grpc_completion_queue_create_for_callback(
+            self._wrapper.c_functor(),
+            NULL
+        )
+
+    cdef grpc_completion_queue* c_ptr(self):
+        return self._cq
+    
+    async def shutdown(self):
+        grpc_completion_queue_shutdown(self._cq)
+        await self._shutdown_completed
+        grpc_completion_queue_destroy(self._cq)
+
+
+async def callback_start_batch(GrpcCallWrapper grpc_call_wrapper,
+                               tuple operations,
+                               object loop):
+    """The callback version of start batch operations."""
+    cdef _BatchOperationTag batch_operation_tag = _BatchOperationTag(None, operations, None)
+    batch_operation_tag.prepare()
+
+    cdef object future = loop.create_future()
+    cdef CallbackWrapper wrapper = CallbackWrapper(
+        future,
+        CallbackFailureHandler('callback_start_batch', operations, RuntimeError))
+    # NOTE(lidiz) Without Py_INCREF, the wrapper object will be destructed
+    # when calling "await". This is an over-optimization by Cython.
+    cpython.Py_INCREF(wrapper)
+    cdef grpc_call_error error = grpc_call_start_batch(
+        grpc_call_wrapper.call,
+        batch_operation_tag.c_ops,
+        batch_operation_tag.c_nops,
+        wrapper.c_functor(), NULL)
+
+    if error != GRPC_CALL_OK:
+        raise RuntimeError("Error with callback_start_batch {}".format(error))
```

Comment:
A reference to the core API is probably a better idea. How about `"grpc_call_start_batch failed {}"`?

Reply:
Good idea!